Null pointer Exception in UnusedLocalVariableCheck

I have read check documentation: https://checkstyle.org/config_coding.html#UnusedLocalVariable
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```
kevin@inspiron-15-5510:~/Desktop/check_style$ cat Test.java
import javax.swing.*;
import java.awt.*;
import java.util.function.Supplier;

public class Test {
   private final NotNullLazyValue<JComponent> myComponent = NotNullLazyValue.lazy(() -> {
      JPanel results = new JPanel(new GridBagLayout()) {};
       return results;
   });
}
class NotNullLazyValue<T> {
    public NotNullLazyValue() {
    }
    public static <T> NotNullLazyValue<T> lazy(Supplier<T> supplier) {
        return new NotNullLazyValue<>();
    }
}

kevin@inspiron-15-5510:~/Desktop/check_style$ javac Test.java

kevin@inspiron-15-5510:~/Desktop/check_style$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN" "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8" />
    <property name="severity" value="warning" />
    <property name="haltOnException" value="false" />

    <module name="TreeWalker">

        <module name="UnusedLocalVariable">
        </module>

    </module>
</module>

kevin@inspiron-15-5510:~/Desktop/check_style$ java -jar /home/kevin/Downloads/checkstyle-10.12.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/kevin/Desktop/check_style/Test.java:1: Got an exception - java.lang.NullPointerException: Cannot invoke "com.puppycrawl.tools.checkstyle.api.DetailAST.getType()" because "parentAst" is null
        at com.puppycrawl.tools.checkstyle.checks.coding.UnusedLocalVariableCheck.getBlockContainingLocalAnonInnerClass(UnusedLocalVariableCheck.java:480)
        at com.puppycrawl.tools.checkstyle.checks.coding.UnusedLocalVariableCheck.visitLocalAnonInnerClass(UnusedLocalVariableCheck.java:404)
        at com.puppycrawl.tools.checkstyle.checks.coding.UnusedLocalVariableCheck.visitToken(UnusedLocalVariableCheck.java:312)
        at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.java:335)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:406)
        at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:273)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:154)
        at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:98)
        at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:334)
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:293)
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:223)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:415)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:338)
        at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:195)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:130)
 [Checker]
Audit done.
Checkstyle ends with 1 errors.
kevin@inspiron-15-5510:~/Desktop/check_style$ 
```
---

# Describe what you expect in detail.
Should Not throw Null pointer Exception

---

--- a/config/pitest-suppressions/pitest-coding-2-suppressions.xml
+++ b/config/pitest-suppressions/pitest-coding-2-suppressions.xml
@@ -1,0 +1,39 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<suppressedMutations>
+  <!-- until https://github.com/checkstyle/checkstyle/issues/14894 -->
+  <mutation unstable="false">
+    <sourceFile>UnusedLocalVariableCheck.java</sourceFile>
+    <mutatedClass>com.puppycrawl.tools.checkstyle.checks.coding.UnusedLocalVariableCheck</mutatedClass>
+    <mutatedMethod>getBlockContainingLocalAnonInnerClass</mutatedMethod>
+    <mutator>org.pitest.mutationtest.engine.gregor.mutators.NegateConditionalsMutator</mutator>
+    <description>negated conditional</description>
+    <lineContent>if (currentAst.getType() == TokenTypes.LAMBDA) {</lineContent>
+  </mutation>
+
+  <mutation unstable="false">
+    <sourceFile>UnusedLocalVariableCheck.java</sourceFile>
+    <mutatedClass>com.puppycrawl.tools.checkstyle.checks.coding.UnusedLocalVariableCheck</mutatedClass>
+    <mutatedMethod>getBlockContainingLocalAnonInnerClass</mutatedMethod>
+    <mutator>org.pitest.mutationtest.engine.gregor.mutators.RemoveConditionalMutator_EQUAL_IF</mutator>
+    <description>removed conditional - replaced equality check with true</description>
+    <lineContent>if (currentAst.getType() == TokenTypes.LAMBDA) {</lineContent>
+  </mutation>
+
+  <mutation unstable="false">
+    <sourceFile>UnusedLocalVariableCheck.java</sourceFile>
+    <mutatedClass>com.puppycrawl.tools.checkstyle.checks.coding.UnusedLocalVariableCheck</mutatedClass>
+    <mutatedMethod>isInsideLocalAnonInnerClass</mutatedMethod>
+    <mutator>org.pitest.mutationtest.engine.gregor.mutators.NegateConditionalsMutator</mutator>
+    <description>negated conditional</description>
+    <lineContent>if (currentAst.getType() == TokenTypes.SLIST) {</lineContent>
+  </mutation>
+
+  <mutation unstable="false">
+    <sourceFile>UnusedLocalVariableCheck.java</sourceFile>
+    <mutatedClass>com.puppycrawl.tools.checkstyle.checks.coding.UnusedLocalVariableCheck</mutatedClass>
+    <mutatedMethod>isInsideLocalAnonInnerClass</mutatedMethod>
+    <mutator>org.pitest.mutationtest.engine.gregor.mutators.RemoveConditionalMutator_EQUAL_IF</mutator>
+    <description>removed conditional - replaced equality check with true</description>
+    <lineContent>if (currentAst.getType() == TokenTypes.SLIST) {</lineContent>
+  </mutation>
+</suppressedMutations>

--- a/config/spotbugs-exclude.xml
+++ b/config/spotbugs-exclude.xml
@@ -315,4 +315,11 @@
     </Or>
     <Bug pattern="CT_CONSTRUCTOR_THROW"/>
   </Match>
+  <Match>
+    <!-- variable is just being saved for further analysis and isn't the main decider
+         of the loop if it is done or not -->
+    <Class name="com.puppycrawl.tools.checkstyle.checks.coding.UnusedLocalVariableCheck"/>
+    <Method name="getBlockContainingLocalAnonInnerClass"/>
+    <Bug pattern="SLS_SUSPICIOUS_LOOP_SEARCH"/>
+  </Match>
 </FindBugsFilter>

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java
@@ -413,15 +413,17 @@ public class UnusedLocalVariableCheck extends AbstractCheck {
     private static DetailAST getBlockContainingLocalAnonInnerClass(DetailAST literalNewAst) {
         DetailAST currentAst = literalNewAst;
         DetailAST result = null;
-        while (!TokenUtil.isOfType(currentAst, CONTAINERS_FOR_ANON_INNERS)) {
-            if (currentAst.getType() == TokenTypes.LAMBDA
-                    && currentAst.getParent()
-                    .getParent().getParent().getType() == TokenTypes.OBJBLOCK) {
-                result = currentAst;
-                break;
+        DetailAST topMostLambdaAst = null;
+        while (currentAst != null && !TokenUtil.isOfType(currentAst, CONTAINERS_FOR_ANON_INNERS)) {
+            if (currentAst.getType() == TokenTypes.LAMBDA) {
+                topMostLambdaAst = currentAst;
             }
             currentAst = currentAst.getParent();
             result = currentAst;
+        }
+
+        if (currentAst == null) {
+            result = topMostLambdaAst;
         }
         return result;
     }

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheckTest.java
@@ -328,6 +328,36 @@ public class UnusedLocalVariableCheckTest extends AbstractModuleTestSupport {
     }
 
     @Test
+    public void testUnusedLocalVarLambdas() throws Exception {
+        final String[] expected = {
+            "14:9: " + getCheckMessage(MSG_UNUSED_LOCAL_VARIABLE, "hoo"),
+            "20:17: " + getCheckMessage(MSG_UNUSED_LOCAL_VARIABLE, "j"),
+            "31:9: " + getCheckMessage(MSG_UNUSED_LOCAL_VARIABLE, "hoo2"),
+            "32:9: " + getCheckMessage(MSG_UNUSED_LOCAL_VARIABLE, "hoo3"),
+            "33:15: " + getCheckMessage(MSG_UNUSED_LOCAL_VARIABLE, "myComponent"),
+            "34:19: " + getCheckMessage(MSG_UNUSED_LOCAL_VARIABLE, "myComponent3"),
+            "40:25: " + getCheckMessage(MSG_UNUSED_LOCAL_VARIABLE, "j"),
+            "52:21: " + getCheckMessage(MSG_UNUSED_LOCAL_VARIABLE, "j"),
+            "65:17: " + getCheckMessage(MSG_UNUSED_LOCAL_VARIABLE, "ja"),
+            "73:9: " + getCheckMessage(MSG_UNUSED_LOCAL_VARIABLE, "k"),
+        };
+        verifyWithInlineConfigParser(
+                getPath("InputUnusedLocalVariableLambdaExpression.java"),
+                expected);
+    }
+
+    @Test
+    public void testUnusedLocalVariableLocalClasses() throws Exception {
+        final String[] expected = {
+            "14:9: " + getCheckMessage(MSG_UNUSED_LOCAL_VARIABLE, "a"),
+            "15:9: " + getCheckMessage(MSG_UNUSED_LOCAL_VARIABLE, "ab"),
+        };
+        verifyWithInlineConfigParser(
+                getPath("InputUnusedLocalVariableLocalClasses.java"),
+                expected);
+    }
+
+    @Test
     public void testUnusedLocalVarRecords() throws Exception {
         final String[] expected = {
             "16:9: " + getCheckMessage(MSG_UNUSED_LOCAL_VARIABLE, "var1"),

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/InputUnusedLocalVariableLambdaExpression.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/InputUnusedLocalVariableLambdaExpression.java
@@ -1,0 +1,92 @@
+/*
+UnusedLocalVariable
+
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.coding.unusedlocalvariable;
+
+import java.util.function.Supplier;
+
+public class InputUnusedLocalVariableLambdaExpression {
+    private final LambdaTest<String> myComponent = LambdaTest.lazy(() -> {
+        String foo = "";
+        String hoo = ""; // violation
+        new Runnable() {
+            String hoo = "";
+
+            @Override
+            public void run() {
+                String j = hoo; // violation
+                String ja = foo;
+                ja += "asd";
+            }
+        };
+        return "";
+    });
+
+    final LambdaTest<String> nestedLambdas = LambdaTest.lazy(() -> {
+        String foo = "";
+        String hoo = "";
+        String hoo2 = ""; // violation
+        String hoo3 = ""; // violation
+        final LambdaTest<String> myComponent = LambdaTest.lazy(() -> { // violation
+            final LambdaTest<String> myComponent3 = LambdaTest.lazy(() -> { // violation
+                new Runnable() {
+                    String hoo2 = "";
+
+                    @Override
+                    public void run() {
+                        String j = hoo; // violation
+                        String ja = foo;
+                        ja += hoo2;
+                    }
+                };
+                return "";
+            });
+            new Runnable() {
+                String hoo3 = "";
+
+                @Override
+                public void run() {
+                    String j = hoo3; // violation
+                    String ja = foo;
+                    ja += "asd";
+                }
+            };
+            return "";
+        });
+        new Runnable() {
+            String hoo2 = "";
+
+            @Override
+            public void run() {
+                String j = hoo2;
+                String ja = foo; // violation
+                j += hoo2;
+            }
+        };
+        return "";
+    });
+
+    final LambdaTest<String> nestedLambdas2 = LambdaTest.lazy(() -> {
+        String k = ""; // violation
+        final LambdaTest<String> nestedLambdas = LambdaTest.lazy(() -> {
+            new LambdaTest<String>() {
+                void foo() {
+                    String j = k;
+                    j += "asd";
+                }
+            };
+            return "";
+        });
+        return nestedLambdas.toString();
+    });
+}
+
+class LambdaTest<T> {
+    String k = "";
+    public static <T> LambdaTest<T> lazy(Supplier<T> supplier) {
+        return new LambdaTest<>();
+    }
+}

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/InputUnusedLocalVariableLocalClasses.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unusedlocalvariable/InputUnusedLocalVariableLocalClasses.java
@@ -1,0 +1,26 @@
+/*
+UnusedLocalVariable
+
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.coding.unusedlocalvariable;
+
+public class InputUnusedLocalVariableLocalClasses {
+
+    int a = 12;
+
+    void foo() {
+        int a = 12; // violation
+        int ab = 1; // violation
+
+        class asd {
+            InputUnusedLocalVariableLocalClasses a = new InputUnusedLocalVariableLocalClasses() {
+                void asd() {
+                    System.out.println(a);
+                }
+            };
+        }
+    }
+
+}

